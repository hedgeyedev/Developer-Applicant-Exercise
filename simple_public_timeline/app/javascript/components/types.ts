export interface User {
  screen_name: string;
  url: string;
  profile_image_url: string;
}

export interface Tweet {
  user: User;
  created_at: string;
  source: string;
  text: string;
  id_str: string;
}
