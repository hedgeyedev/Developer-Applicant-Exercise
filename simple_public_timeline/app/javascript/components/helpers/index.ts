export const formatDateString = (date: string) => {
  return new Date(date).toLocaleTimeString('en-us', { hour: 'numeric', minute: '2-digit' ,timeZone: 'UTC'});
};
