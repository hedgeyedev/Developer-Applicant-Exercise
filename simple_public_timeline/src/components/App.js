import React from 'react';


const axios = require('axios');

const App = () => {
  return (
    <div>
      <h1>Recent Public Tweets</h1>
      <h2>What everyone on Twitter is talking about</h2>
      {axios.get('/tweets')
        .then(function (response) {
          response.list.map(tweet =>
            <div class="card">
              <div class="container">
                <h4><b>tweet.username</b></h4>
                <p>tweet.text</p>
                <footer>tweet.created_at</footer>
              </div>
            </div>
          )
        })
        .catch(function (error) {
          console.log(error);
        })}
    </div>
  );
};

export default App;
