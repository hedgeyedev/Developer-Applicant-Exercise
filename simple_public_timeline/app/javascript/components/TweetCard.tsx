import React from "react";
import ReactHtmlParser from "react-html-parser";
import { formatDateString } from "./helpers";
import { Tweet } from "./types";

interface TweetCardProps {
  tweet: Tweet;
}

/**
 * Renders a tweet
 * @param Tweet a tweet with profile image, user, and created date
 */
export const TweetCard = ({ tweet }: TweetCardProps) => {
  const {
    user: { url, screen_name, profile_image_url },
    created_at,
    source,
    text,
  } = tweet;

  return (
    <div className="tweet-card">
      <div className="icon">
        <img src={profile_image_url} />
      </div>
      <div className="tweet-body">
        <p>
          <a href={url}>{screen_name}</a>
          {text}
        </p>
        <h5>{formatDateString(created_at)}</h5>
        <h6>{ReactHtmlParser(source)}</h6>
      </div>
    </div>
  );
};
