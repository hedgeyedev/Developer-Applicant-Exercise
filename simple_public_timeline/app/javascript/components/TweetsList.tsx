import React, { useEffect, useState } from "react";
import ReactHtmlParser from "react-html-parser";
import { formatDateString } from "./helpers";

interface TweetsListProps {
  tweets_url: string;
}

interface User {
  screen_name: string;
  url: string;
  profile_image_url: string;
}

interface Tweet {
  user: User;
  created_at: string;
  source: string;
  text: string;
  id_str: string;
}

interface TweetCardProps {
  tweet: Tweet;
}

/**
 * Renders a tweet
 * @param Tweet a tweet with profile image, user, and created date
 */
export const TweetCard = ({ tweet }: TweetCardProps) => {
  const {
    user: { url, screen_name, profile_image_url },
    created_at,
    source,
    text,
  } = tweet;

  return (
    <div className="tweet-card">
      <div className="icon">
        <img src={profile_image_url} />
      </div>
      <div className="tweet-body">
        <p>
          <a href={url}>{screen_name}</a>
          {text}
        </p>
        <h5>{formatDateString(created_at)}</h5>
        <h6>{ReactHtmlParser(source)}</h6>
      </div>
    </div>
  );
};

/**
 * Renders a list of tweets as cards
 *
 * @param tweets_url Url to fetch the set of tweets. In this case it is passed as
 * The rails helper to via_js
 */
const TweetsList = ({ tweets_url }: TweetsListProps) => {
  const [recentTweets, setRecentTweets] = useState<Tweet[]>([]);

  // Runs when the component mounts. We return an empty array from useEffect
  // so that this only runs once after mount. If we were going to poll for updates
  // or use server sent events, we would set that up here. Here we just make
  // a basic GET request and update the state with the tweets.
  useEffect(() => {
    (async () => {
      // Make the request for JSON
      const response = await fetch(`${tweets_url}`, {
        headers: { contentType: "application/json" },
      });

      if (!response.ok) {
        // Handle any errors here. Something like a toast
        return;
      }
      // get the response JSON
      const tweets: Tweet[] = await response.json();
      console.log(tweets);
      // Update the state with the tweets
      setRecentTweets(tweets);
    })();
  }, []);

  return (
    <div>
      {recentTweets.map((tweet) => {
        return <TweetCard key={tweet.id_str} tweet={tweet} />;
      })}
    </div>
  );
};

export default TweetsList;
