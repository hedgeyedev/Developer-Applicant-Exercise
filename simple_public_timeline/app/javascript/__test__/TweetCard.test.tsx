import * as React from "react";
import { TweetCard } from "../components/TweetsList";
import * as renderer from "react-test-renderer";

it("render correctly text component", () => {
  const props = {
    tweet: {
      user: {
        screen_name: "John",
        url: "http://myprofile",
        profile_image_url: "string",
      },
      created_at: "Sun Mar 21 22:05:09 +0000 2021",
      source:
        '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>',
      text: "some cool tweet test text",
      id_str: "1373757599909224449",
    },
  };
  const TextInputComponent = renderer.create(<TweetCard {...props} />).toJSON();
  expect(TextInputComponent).toMatchSnapshot();
});
