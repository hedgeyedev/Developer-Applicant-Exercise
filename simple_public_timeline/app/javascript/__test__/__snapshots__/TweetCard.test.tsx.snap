// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly text component 1`] = `
<div
  className="tweet-card"
>
  <div
    className="icon"
  >
    <img
      src="string"
    />
  </div>
  <div
    className="tweet-body"
  >
    <p>
      <a
        href="http://myprofile"
      >
        John
      </a>
      some cool tweet test text
    </p>
    <h5>
      10:05 PM
    </h5>
    <h6>
      <a
        href="http://twitter.com/download/android"
        rel="nofollow"
      >
        Twitter for Android
      </a>
    </h6>
  </div>
</div>
`;
